src/main/java/net/dv8tion/jda/api/entities/Template.java

```diff
@@ -0,0 +1,388 @@
+     */
+    @Nonnull
+    String getCode();
+
+    /**
+     * The template name.
+     *
+     * @return The template name
+     */
+    @Nonnull
+    String getName();
+
+    /**
+     * The template description.
+     *
+     * @return The template description
+     */
+    @Nullable
+    String getDescription();
+
+    /**
+     * How often this template has been used.
+     *
+     * @return The uses of this template
+     */
+    int getUses();
+
+    /**
+     * The user who created this template.
+     *
+     * @return The user who created this template
+     */
+    @Nonnull
+    User getCreator();
+
+    /**
+     * Returns creation date of this template.
+     *
+     * @return The creation date of this template
+     */
+    @Nonnull
+    OffsetDateTime getTimeCreated();
+
+    /**
+     * Returns the last update date of this template.
+     * <br>If this template has never been updated, this returns the date of creation.
+     *
+     * @return The last update date of this template
+     *
+     * @see    #getTimeCreated()
+     */
+    @Nonnull
+    OffsetDateTime getTimeUpdated();
+
+    /**
+     * A {@link net.dv8tion.jda.api.entities.Template.Guild Template.Guild} object
+     * containing information about this template's origin guild.
+     *
+     * @return Information about this template's origin guild
+     *
+     * @see    net.dv8tion.jda.api.entities.Template.Guild
+     */
+    @Nonnull
+    Guild getGuild();
+
+    /**
+     * Whether or not this template is synced.
+     *
+     * @return True, if this template matches the current guild structure
+     */
+    boolean isSynced();
+
+    /**
+     * Returns the {@link net.dv8tion.jda.api.managers.TemplateManager TemplateManager} for this Template.
+     * <br>In the TemplateManager, you can modify the name or description of the template.
+     * You modify multiple fields in one request by chaining setters before calling {@link net.dv8tion.jda.api.requests.RestAction#queue() RestAction.queue()}.
+     *
+     * <p>This is a lazy idempotent getter. The manager is retained after the first call.
+     * This getter is not thread-safe and would require guards by the user.
+     *
+     * @throws IllegalStateException
+     *         if the account is not in the template's guild
+     * @throws net.dv8tion.jda.api.exceptions.InsufficientPermissionException
+     *         If the currently logged in account does not have {@link net.dv8tion.jda.api.Permission#MANAGE_SERVER MANAGE_SERVER}
+     *
+     * @return The TemplateManager of this Template
+     */
+    @Nonnull
+    TemplateManager getManager();
+
+    /**
+     * The {@link net.dv8tion.jda.api.JDA JDA} instance used to create this Template instance.
+     *
+     * @return The corresponding JDA instance
+     */
+    @Nonnull
+    JDA getJDA();
+
+    /**
+     * POJO for the guild information provided by a template.
+     *
+     * @see #getGuild()
+     */
+    interface Guild extends ISnowflake
```

Comment:
What about channels?

Reply:
Considering how many nested classes this has, maybe it makes more sense to just make a new package for this?

```
entities.templates.Template
entities.templates.TemplateGuild
entities.templates.TemplateRole
entities.templates.TemplateChannel
...
```

Reply:
I also think it might make sense to just make these directly as classes rather than interface+impl. These are immutable data holders anyway, so no need for hiding setters.

Reply:
i asked about how to implement channels before [here](https://github.com/DV8FromTheWorld/JDA/pull/1465/#discussion_r546263769), but got no response, so i didn't implement it as i'm not sure how to implement multiple types of channels

Reply:
By using the ChannelType enum? You don't need to make a different class for voice/text/news etc. Just make things that are only available for text channels nullable, etc.